File: frontend/services/auth_service.py
"""
Service d'authentification pour gérer les tokens JWT et l'état utilisateur
"""
import streamlit as st
import requests
import json
from typing import Optional, Dict

API_URL = "http://localhost:8000"

class AuthService:
    """Service pour gérer l'authentification des utilisateurs"""
    
    @staticmethod
    def login(username: str, password: str) -> Dict:
        """Connexion utilisateur"""
        try:
            response = requests.post(
                f"{API_URL}/api/auth/login",
                data={"username": username, "password": password}
            )
            
            if response.status_code == 200:
                data = response.json()
                # Sauvegarder dans session_state
                st.session_state.token = data["access_token"]
                st.session_state.user = data["user"]
                st.session_state.is_authenticated = True
                return {"success": True, "data": data}
            else:
                return {"success": False, "error": response.json().get("detail", "Erreur de connexion")}
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def register(username: str, email: str, password: str) -> Dict:
        """Inscription utilisateur"""
        try:
            response = requests.post(
                f"{API_URL}/api/auth/register",
                json={"username": username, "email": email, "password": password}
            )
            
            if response.status_code in [200, 201]:
                return {"success": True, "data": response.json()}
            else:
                return {"success": False, "error": response.json().get("detail", "Erreur d'inscription")}
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    @staticmethod
    def logout():
        """Déconnexion utilisateur"""
        # Nettoyer session_state
        if 'token' in st.session_state:
            del st.session_state.token
        if 'user' in st.session_state:
            del st.session_state.user
        if 'is_authenticated' in st.session_state:
            del st.session_state.is_authenticated
        st.rerun()
    
    @staticmethod
    def is_authenticated() -> bool:
        """Vérifier si l'utilisateur est authentifié"""
        return st.session_state.get('is_authenticated', False)
    
    @staticmethod
    def get_token() -> Optional[str]:
        """Récupérer le token JWT"""
        return st.session_state.get('token', None)
    
    @staticmethod
    def get_user() -> Optional[Dict]:
        """Récupérer les informations utilisateur"""
        return st.session_state.get('user', None)
    
    @staticmethod
    def get_headers() -> Dict:
        """Récupérer les headers avec le token d'authentification"""
        token = AuthService.get_token()
        if token:
            return {"Authorization": f"Bearer {token}"}
        return {}

File: frontend/src/services/authService.js
/**
 * Service d'authentification
 * Gère le login, register et la persistance du token
 */

import axios from 'axios';

const API_URL = 'http://localhost:8000/api';

// Créer une instance axios
const api = axios.create({
  baseURL: API_URL,
});

// Intercepteur pour ajouter le token à chaque requête
api.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('token');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// Intercepteur pour gérer les erreurs 401 (non authentifié)
api.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response?.status === 401) {
      // Token expiré ou invalide
      localStorage.removeItem('token');
      localStorage.removeItem('user');
      window.location.href = '/login';
    }
    return Promise.reject(error);
  }
);

const authService = {
  // Inscription
  async register(userData) {
    const response = await axios.post(`${API_URL}/auth/register`, userData);
    if (response.data.access_token) {
      localStorage.setItem('token', response.data.access_token);
      localStorage.setItem('user', JSON.stringify(response.data.user));
      // Déclencher un événement pour notifier l'application
      window.dispatchEvent(new Event('auth-change'));
    }
    return response.data;
  },

  // Connexion
  async login(username, password) {
    const formData = new FormData();
    formData.append('username', username);
    formData.append('password', password);
    
    const response = await axios.post(`${API_URL}/auth/login`, formData);
    if (response.data.access_token) {
      localStorage.setItem('token', response.data.access_token);
      localStorage.setItem('user', JSON.stringify(response.data.user));
      // Déclencher un événement pour notifier l'application
      window.dispatchEvent(new Event('auth-change'));
    }
    return response.data;
  },

  // Déconnexion
  logout() {
    localStorage.removeItem('token');
    localStorage.removeItem('user');
    window.location.href = '/login';
  },

  // Vérifier si l'utilisateur est connecté
  isAuthenticated() {
    try {
      const token = localStorage.getItem('token');
      return !!token;
    } catch (error) {
      console.error('Erreur lors de la vérification du token:', error);
      return false;
    }
  },

  // Récupérer l'utilisateur connecté
  getCurrentUser() {
    try {
      const userStr = localStorage.getItem('user');
      return userStr ? JSON.parse(userStr) : null;
    } catch (error) {
      console.error('Erreur lors de la récupération de l\'utilisateur:', error);
      return null;
    }
  },

  // Récupérer le token
  getToken() {
    try {
      return localStorage.getItem('token');
    } catch (error) {
      console.error('Erreur lors de la récupération du token:', error);
      return null;
    }
  },

  // Vérifier l'authentification avec le serveur
  async checkAuth() {
    try {
      const response = await api.get('/auth/check');
      return response.data;
    } catch (error) {
      return null;
    }
  }
};

export { api };
export default authService;
